string_utils/fms_string_utils.F90

```diff
@@ -248,30 +240,52 @@ subroutine fms_f2c_string (dest, str_in)
     enddo
 end subroutine fms_f2c_string
 
-  !> @brief Converts a Boolean value to a string
-  !> @return The Boolean value as a string
-  function string_from_logical(v)
-    logical, intent(in) :: v !< Boolean value to be converted to a string
-    character(:), allocatable :: string_from_logical
-
-    if (v) then
-      string_from_logical = "True"
-    else
-      string_from_logical = "False"
-    endif
+  !> @brief Converts a number or a Boolean value to a string
+  !> @return The argument as a string
+  function string(v, fmt)
+    class(*), intent(in) :: v ! Value to be converted to a string
+    character(*), intent(in), optional :: fmt !< Optional format string for a real argument
+    character(:), allocatable :: string
+
+    select type(v)
+      type is (logical)
+        if (v) then
+          string = "True"
+        else
+          string = "False"
+        endif
+
+      type is (integer)
+        allocate(character(32) :: string)
+        write(string, '(i0)') v
+        string = trim(adjustl(string))
+
+      type is (real(r4_kind))
+        allocate(character(32) :: string)
+        if (present(fmt)) then
+          write(string, "(" // fmt // ")") v
+        else
+          write(string, *) v
+        endif
+        string = trim(adjustl(string))
+
+      type is (real(r8_kind))
+        allocate(character(32) :: string)
+        if (present(fmt)) then
+          write(string, "(" // fmt // ")") v
+        else
+          write(string, *) v
+        endif
+        string = trim(adjustl(string))
+
+      type is (character(*))
+        string = v
+
+      class default
+        call mpp_error(FATAL, "string() called with incompatible argument")
```

Comment:
If you keep the error message, it should be more comprehensive.  `"string() called with incompatible argument.  Must be of type integer(kind=4), integer(kind=8), real(kind=4), real(kind=8), or logical."`